tsdb/engine/tsm1/tombstone.go

```diff
@@ -323,6 +338,70 @@ func (t *Tombstoner) readTombstoneV2(f *os.File, fn func(t Tombstone) error) err
 	}
 }
 
+// readTombstoneV3 reads the third version of tombstone files that are capable
+// of storing keys and the range of time for the key that points were deleted. This
+// format is a binary and compressed with gzip.
+func (t *Tombstoner) readTombstoneV3(f *os.File, fn func(t Tombstone) error) error {
+	// Skip header, already checked earlier
+	if _, err := f.Seek(headerSize, io.SeekStart); err != nil {
+		return err
+	}
+	n := int64(headerSize)
+
+	var (
+		min, max int64
+		key      string
+	)
+
+	gr, err := gzip.NewReader(bufio.NewReader(f))
+	if err != nil {
+		return err
+	}
+	defer gr.Close()
+
+	b := make([]byte, 4096)
+	for {
+		if _, err = io.ReadFull(gr, b[:4]); err == io.EOF || err == io.ErrUnexpectedEOF {
+			return nil
+		} else if err != nil {
+			return err
+		}
+		n += 4
+
+		keyLen := int(binary.BigEndian.Uint32(b[:4]))
```

Comment:
Paranoid me... But.. On a 32-bit system isn't it going to be possible for the `keyLen` to be a value larger than `math.MaxInt32`, and therefore when we convert to an `int` (which will be an `int32` on a 32-bit system), we may overflow.

Probably unlikely, but it could be a bit safer to do: 

```go
keyLen := binary.BigEndian.Uint32(b[:4])
if keyLen > uint32(len(b)) {
...
}
```

Just a thought.